Assets/MRTK/SDK/Experimental/Dock/Dockable.cs

```diff
@@ -0,0 +1,335 @@
+                "A Dockable object must have either an ObjectManipulator or a ManipulationHandler component.");
+        }
+
+        /// <summary>
+        /// Unsubscribes from manipulation events.
+        /// </summary>
+        private void OnDestroy()
+        {
+            if (objectManipulator != null)
+            {
+                objectManipulator.OnManipulationStarted.RemoveListener(OnManipulationStarted);
+                objectManipulator.OnManipulationEnded.RemoveListener(OnManipulationEnded);
+
+                objectManipulator = null;
+            }
+
+            if (manipulationHandler != null)
+            {
+                manipulationHandler.OnManipulationStarted.RemoveListener(OnManipulationStarted);
+                manipulationHandler.OnManipulationEnded.RemoveListener(OnManipulationEnded);
+
+                manipulationHandler = null;
+            }
+
+            if (dockedPosition != null)
+            {
+                dockedPosition.dockedObject = null;
+                dockedPosition = null;
+            }
+
+            overlappingPositions.Clear();
+        }
+
+        /// <summary>
+        /// Updates the transform and state of this object every frame, depending on 
+        /// manipulations and docking state.
+        /// </summary>
+        public void Update()
+        {
+            if (isDragging && overlappingPositions.Count > 0)
+            {
+                var closestPosition = GetClosestPosition();
+                if (closestPosition.IsOccupied)
+                {
+                    closestPosition.GetComponentInParent<Dock>().TryMoveToFreeSpace(closestPosition);
+                }
+            }
+
+            if (dockingState == DockingState.Docked || dockingState == DockingState.Docking)
+            {
+                Assert.IsNotNull(dockedPosition, "When a dockable is docked, its dockedPosition must be valid.");
+                Assert.AreEqual(dockedPosition.dockedObject, this, "When a dockable is docked, its dockedPosition reference the dockable.");
+
+                var lerpTime = dockingState == DockingState.Docked ? moveLerpTimeWhenDocked : moveLerpTime;
+
+                if (!isDragging)
+                {
+                    // Don't override dragging
+                    transform.position = Solver.SmoothTo(transform.position, dockedPosition.transform.position, Time.deltaTime, lerpTime);
+                    transform.rotation = Solver.SmoothTo(transform.rotation, dockedPosition.transform.rotation, Time.deltaTime, lerpTime);
+                }
+
+                transform.localScale = Solver.SmoothTo(transform.localScale, dockedPositionScale, Time.deltaTime, lerpTime);
+
+                if (CloseEnough(dockedPosition.transform.position, transform.position) &&
+                    AlignedEnough(dockedPosition.transform.rotation, transform.rotation) &&
+                    AboutTheSameSize(dockedPositionScale, transform.localScale))
```

Comment:
I'd consider moving these helpers to extension methods. `VectorExtensions` and `QuaternionExtensions`.